### PR TITLE
feat(webhooks): add support for custom outgoing webhook http headers

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This pr adds support for custom outgoing webhook http headers. If we pass `custom_outgoing_webhook_http_headers` to business profile these headers will be forwaded to all outgoing webhooks request headers. 

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Test case 1 : 
Create or Update the business profile with custom http headers
```
curl --location 'http://localhost:8080/account/merchant_1720605139/business_profile/pro_vGz02PKyssUnujEpTwee' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
  "profile_name": "food",
  "outgoing_webhook_custom_http_headers": {
    "authentication" : "this is test key header",
    "client_id" : "askdjfhaksjhdf",
    "client_key" : "asdkfjhakdjh"
  }

}'
```

Create payment you should should be able to see custom headers in the configure test webhook site.
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_WYTknsPs0yvmcht1il3zk5bwvieOxmpk43I1O17xUrGDMgFoKYqUErBeafZw43Tg' \
--data-raw '{
    "amount":1000,
    "currency": "EUR",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2029-09-10T10:11:12Z",
    "amount_to_capture": 1000,
    "customer_id": "AdyenCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method_data": {
        "card": {
            "card_number": "4111111145551142",
            "card_exp_month": "03",
            "card_exp_year": "30",
            "card_holder_name": "chikke.srujan@juspay.in",
            "card_cvc": "737",
            "card_network": "Visa",
            "card_type": null,
            "card_issuing_country": null,
            "bank_code": null
        },
        "billing": {
            "address": {
                "line1": "1467",
                "line2": "CA",
                "line3": "CA",
                "city": "San Fransico",
                "state": "California",
                "zip": "94122",
                "country": "NL",
                "first_name": "joseph",
                "last_name": "Doe"
            },
            "phone": {
                "number": "8056594427",
                "country_code": "+91"
            }
        }
    },
    "payment_method": "card",
    "payment_method_type" :"credit",
    "customer_acceptance": {
        "acceptance_type": "online",
        "accepted_at": "2026-07-04T10:59:42.232Z",
        "online": {
            "ip_address": null,
            "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0"
        }
    },
    "browser_info": {
        "color_depth": 24,
        "java_enabled": true,
        "java_script_enabled": true,
        "language": "en-GB",
        "screen_height": 720,
        "screen_width": 1280,
        "time_zone": -330,
        "ip_address": "208.127.127.193",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "CA",
            "line3": "CA",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
Screenshot 2024-07-10 at 5 45 10 PM

Test case 2 : Pass random string to the field `custom_outgoing_webhook_http_headers` , this should not trigger custom headers to outgoing webhook, but this will show warning in the logs 
Screenshot 2024-07-10 at 5 55 12 PM

Testcase 3 : Check how data is stored in db, `outgoing_webhook_custom_http_headers` should be encrypted in db.

Screenshot 2024-07-15 at 7 34 43 PM

Test case 4: check whether retry for outgoing webhooks have custom headers . You can verify it by passing invalid return url to create 4xx webhook. 
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/webhooks/outgoing.rs [hunk 1]
```diff
@@ -549,20 +548,36 @@ fn get_webhook_url_from_business_profile(
...
     ) -> CustomResult<OutgoingWebhookRequestContent, errors::WebhooksFlowError> {
         let mut headers = vec![(
             reqwest::header::CONTENT_TYPE.to_string(),
             mime::APPLICATION_JSON.essence_str().into(),
         )];
 
         let transformed_outgoing_webhook = WebhookType::from(outgoing_webhook);
-
+        let payment_response_hash_key = business_profile.payment_response_hash_key.clone();
+        let hashmap = business_profile
```
Comment (SanchithHegde) [inline crates/router/src/core/webhooks/outgoing.rs:565]: Nit: naming.

```suggestion
        let custom_headers = business_profile
```

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/core/admin.rs [hunk 2]
```diff
@@ -1888,8 +1897,9 @@ pub async fn create_business_profile(
     }
 
     Ok(service_api::ApplicationResponse::Json(
-        api_models::admin::BusinessProfileResponse::foreign_try_from(business_profile)
-            .change_context(errors::ApiErrorResponse::InternalServerError)?,
+        admin::business_profile_response(business_profile, &key_store)
+            .change_context(errors::ApiErrorResponse::InternalServerError)
```
Comment (SanchithHegde) [inline crates/router/src/core/admin.rs:1901]: Can you add a context message with `attach_printable` or `attach_printable_lazy`?

File: crates/router/src/core/admin.rs [hunk 3]
```diff
@@ -1926,8 +1947,9 @@ pub async fn retrieve_business_profile(
         })?;
 
     Ok(service_api::ApplicationResponse::Json(
-        api_models::admin::BusinessProfileResponse::foreign_try_from(business_profile)
-            .change_context(errors::ApiErrorResponse::InternalServerError)?,
+        admin::business_profile_response(business_profile, &key_store)
+            .change_context(errors::ApiErrorResponse::InternalServerError)
```
Comment (SanchithHegde) [inline crates/router/src/core/admin.rs:1951]: Add context message.

File: crates/router/src/core/admin.rs [hunk 4]
```diff
@@ -2075,8 +2113,9 @@ pub async fn update_business_profile(
         })?;
 
     Ok(service_api::ApplicationResponse::Json(
-        api_models::admin::BusinessProfileResponse::foreign_try_from(updated_business_profile)
-            .change_context(errors::ApiErrorResponse::InternalServerError)?,
+        admin::business_profile_response(updated_business_profile, &key_store)
+            .change_context(errors::ApiErrorResponse::InternalServerError)
```
Comment (SanchithHegde) [inline crates/router/src/core/admin.rs:2117]: Add context message.

File: crates/router/src/core/webhooks/outgoing.rs [hunk 5]
```diff
@@ -86,10 +89,12 @@ pub(crate) async fn create_event_and_trigger_outgoing_webhook(
...
         &merchant_account,
         outgoing_webhook,
-        business_profile.payment_response_hash_key.as_deref(),
+        &business_profile,
+        merchant_key_store,
     )
+    .await
     .change_context(errors::ApiErrorResponse::WebhookProcessingFailure)
-    .attach_printable("Failed to construct outgoing webhook request content")?;
+    .attach_printable("Failed to construct outgoing webhook custom HTTP headers")?;
```
Comment (SanchithHegde) [inline crates/router/src/core/webhooks/outgoing.rs:97]: Can you revert this change?

File: crates/router/src/core/webhooks/outgoing.rs [hunk 6]
```diff
@@ -546,23 +551,49 @@ fn get_webhook_url_from_business_profile(
...
 
         let transformed_outgoing_webhook = WebhookType::from(outgoing_webhook);
-
+        let payment_response_hash_key = business_profile.payment_response_hash_key.clone();
+        let custom_headers = decrypt::<serde_json::Value, masking::WithType>(
+            business_profile
+                .outgoing_webhook_custom_http_headers
+                .clone(),
+            key_store.key.get_inner().peek(),
+        )
```
Comment (SanchithHegde) [inline crates/router/src/core/webhooks/outgoing.rs:578]: We'd preferably want to have this implementation as part of domain type implementation rather than encrypting/decrypting it individually in separate locations.

File: crates/router/src/types/api/admin.rs [hunk 7]
```diff
@@ -80,169 +83,189 @@ impl ForeignTryFrom<domain::MerchantAccount> for MerchantAccountResponse {
...
+        payment_response_hash_key: Some(payment_response_hash_key),
+        redirect_to_merchant_with_http_post: request
+            .redirect_to_merchant_with_http_post
+            .unwrap_or(merchant_account.redirect_to_merchant_with_http_post),
+        webhook_details: webhook_details.or(merchant_account.webhook_details),
+        metadata: request.metadata,
+        routing_algorithm: Some(serde_json::json!({
+            "algorithm_id": null,
+            "timestamp": 0
+        })),
```
Comment (SanchithHegde) [inline crates/router/src/types/api/admin.rs:219]: Do we still need this hardcoded value?

Comment (srujanchikke) [reply crates/router/src/types/api/admin.rs:219]: @SanchithHegde this was hardcoded for backward compatibility. 
cc: @Aprabhat19

File: crates/router/src/core/webhooks/outgoing.rs (hunk 6 above)
Comment (srujanchikke) [reply crates/router/src/core/webhooks/outgoing.rs:578]: Sure will take this in separate pr https://github.com/juspay/hyperswitch-cloud/issues/5984

**Review by SanchithHegde — APPROVED** [review]

**Review by sai-harsha-vardhan — DISMISSED** [review]

**Review by Aprabhat19 — DISMISSED** [review]

**Review by Aprabhat19 — DISMISSED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]
